File: src/org/newtonpolyhedron/solve/poly/PolyMotzkinBurgerSolver.java
package org.newtonpolyhedron.solve.poly;

import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;

import org.fs.utils.collection.table.ArrayListKeyTable;
import org.fs.utils.collection.table.KeyTable;
import org.fs.utils.collection.table.KeyTables;
import org.newtonpolyhedron.entity.vector.FractionVector;
import org.newtonpolyhedron.entity.vector.IntVector;
import org.newtonpolyhedron.solve.cone.ConeSolver;
import org.newtonpolyhedron.utils.NullPrintWriter;
import org.newtonpolyhedron.utils.PointUtils;

/** @author Alexander "FS" Abdugafarov */
public class PolyMotzkinBurgerSolver implements PolyhedronSolver {
	
	private final ConeSolver	coneSolver;
	
	public PolyMotzkinBurgerSolver(final ConeSolver coneSolver) {
		this.coneSolver = coneSolver;
	}
	
	@Override
	public KeyTable <IntVector, Integer, Boolean> solve(
			final List <FractionVector> points,
			final List <IntVector> commonLimits,
			final List <IntVector> wishfulBasis,
			final PrintWriter output) throws Exception {
		final int dim = points.get(0).getDim();
		
		final KeyTable <IntVector, Integer, Boolean> lookupTable = new ArrayListKeyTable <IntVector, Integer, Boolean>();
		fillTableIdxKeys(lookupTable, points.size());
		
		final List <IntVector> normalVectorList = new ArrayList <IntVector>();
		
		for (int currPtIdx = 0; currPtIdx < points.size(); ++currPtIdx) {
			// Forming a list of equations by substracting current point
			final List <IntVector> pointless = PointUtils.copySubtractPointAsInt(points, currPtIdx);
			
			// Adding a common limits - if any
			if (commonLimits != null) {
				pointless.addAll(commonLimits);
			}
			
			final List <IntVector> tempSols = coneSolver.solve(pointless, wishfulBasis, dim, NullPrintWriter.instance);
			
			for (final IntVector tempSol : tempSols) {
				boolean alreadyThere = false;
				for (final IntVector normalVec : normalVectorList) {
					if (normalVec.equals(tempSol)) {
						lookupTable.put(normalVec, currPtIdx, true);
						alreadyThere = true;
						break;
					}
				}
				if (!alreadyThere) {
					normalVectorList.add(tempSol);
					lookupTable.put(tempSol, currPtIdx, true);
				}
			}
		}
		
		// lookupTableSort(normalVectorList, lookupTable, false);
		KeyTables.sortByRowHeaders(lookupTable, true);
		KeyTables.sortByColHeaders(lookupTable, true);
		
		return lookupTable;
	}
	
	private void fillTableIdxKeys(final KeyTable <IntVector, Integer, Boolean> lookupTable, final int upTo) {
		for (int i = 0; i < upTo; ++i) {
			lookupTable.put(null, i, null);
		}
		lookupTable.removeRow(null);
	}
}

File: src/main/scala/org/newtonpolyhedron/solve/poly/PolyMotzkinBurgerSolver.scala
package org.newtonpolyhedron.solve.poly

import java.io.PrintWriter
import java.util.Comparator

import org.fs.utility.collection.table.KeyTable
import org.newtonpolyhedron.NewtonImports._
import org.newtonpolyhedron.solve.cone.ConeSolver
import org.newtonpolyhedron.utils.PointUtils

class PolyMotzkinBurgerSolver[N <: MPNumber](val coneSolver: ConeSolver)(implicit mp: MathProcessor[N])
    extends PolyhedronSolver[N] {

  override def solve(
      points:             Seq[NumVec[N]],
      commonLimitsOption: Option[Seq[IntVec]],
      wishfulBasisOption: Option[Seq[IntVec]],
      output:             PrintWriter
  ): KeyTable[IntVec, Int, Boolean] = {
    val allSolutions = solveForEachPoint(points, commonLimitsOption, wishfulBasisOption, output)
    fillTableWith(allSolutions)
  }

  def solveForEachPoint(
      points:             Seq[NumVec[N]],
      commonLimitsOption: Option[Seq[IntVec]],
      wishfulBasisOption: Option[Seq[IntVec]],
      output:             PrintWriter
  ): Seq[Seq[IntVec]] = {
    val dim = points.head.size
    val allSolutions = for (currPtIdx <- 0 until points.size) yield {
      val commonLimits = commonLimitsOption getOrElse IndexedSeq.empty
      // Forming equations by substracting current point, plus common limits - if any
      val eqSys = PointUtils.copySubtractPointAsInt(points, currPtIdx) ++ commonLimits
      val coneSolutions = coneSolver.solve(eqSys, wishfulBasisOption, dim)
      coneSolutions
    }
    allSolutions
  }

  val intComparator: Comparator[Int] = new Comparator[Int] {
    override def compare(o1: Int, o2: Int) = o1 compare o2
  }

  def fillTableWith(allSolutions: Seq[Seq[IntVec]]): KeyTable[IntVec, Int, Boolean] = {
    // Looks way cleaner with variable than with values and folds
    var lookupTable = KeyTable.empty[IntVec, Int, Boolean]
    for (i <- 0 until allSolutions.size)
      lookupTable = lookupTable.withEmptyCol(i)
    for {
      (coneSols, i) <- allSolutions.zipWithIndex
      sol <- coneSols
    } lookupTable = lookupTable + (sol, i, true)
    lookupTable = lookupTable.sortedCols
    lookupTable = lookupTable.sortedRows
    lookupTable
  }
}
